### DISCUSSION TITLE
High availability?

### DISCUSSION THREAD

**Discussion by fkleedorfer** [eclipse-rdf4j/rdf4j#5147] [Q&A]
There is a [HA solution for Jena](https://afs.github.io/rdf-delta/ha-system.html) Is there anything comparable for RDF4J? 

(Alternatively, is there appetite for creating it?)

**Comment by kenwenzel** [comment]
Not yet, but I would propose to create something based on Apache Ratis.

**Reply by fkleedorfer** [reply]
Very interesting. Want to start working on a prototype?

**Reply by kenwenzel** [reply]
Not directly, but it is somehow on my to-do list.
We could start with the FileStore example and use the NativeStore or LmdbStore for persistent storage.
I'm not yet sure if we should support distributed transactions or if it is enough to rely on SnapshotSailStore and only replicate committed changes via RAFT. The latter would only work if we have one master and multiple replicas.
A real distributed store would also be interesting.

**Reply by fkleedorfer** [reply]
By distributed transactions you mean that multiple server instances participate in one transaction? Not sure about the cost benefit ratio there.

How about the next easier thing: have a high throughput, data-less rdf4j server as a load balancer that routes to the HA instances. Open transactions are always routed to the same instance. We replicate writes to all HA instances, which will be consistent eventually. The hard part would be to avoid read-read,-write-write inconsistencies, which maybe we can do by causing the raft algo to refuse the second commit.

Then there is the question how we would replicate a transaction. Is there a way to obtain added and removed quads in a commit? If so, (even for "update where" queries) that's the thing we should use.

**Reply by fkleedorfer** [reply]
Also, we should allow SHACL validation in the original commit but suppress it on replicated commits

**Reply by hmottestad** [reply]
Leader election can be very tricky. Might be easier to build a solution with a single master node and replicated read nodes.

Clients can decide if they want to read from a quorum or just a single node. If reading from a single node they could be reading stale data. Reads from quorum would not actually read the data from multiple nodes, just have a vote to see which snapshot is the latest and then read that from a single node. Writes would always have to be quorum, as in that a write is only complete once it's been replicated to more than half the read replicas. 

Raft would be good to keep track of which nodes are part of the cluster. Once a node is kicked out of the cluster it can not rejoin before the cluster is restarted.

This would support:
 - single writer node
 - multiple read nodes
 - redundancy for reads, as long as more than half the nodes are up
 - no redundancy for writes, once the write node is down the entire cluster would need to be restarted

**Comment by hmottestad** [comment]
I would like to start another thread here to talk about implementing support for Jena's RDF Delta Patch Log Server instead of actually developing a HA system from scratch. 

What makes the Jena setup HA is that the patch log server runs in HA with multiple servers. 

We could create a Sail that interfaces with the patch log server to track which patch version we are on, fetch any newer patches and apply them to the underlying store, and also sync back changes to the patch log server before a transaction is allowed to finish committing. 

The NotifyingSail would allow us to track changes in a transaction, and we can override the begin, prepare and commit methods to add the interaction with the patch log server. 

An important simplification that the patch log server allows for is that a transaction starts at a patch version, and needs to be commit at the same version. This means that a high replicated write load would end up with a lot of cancelled transactions.

**Reply by kenwenzel** [reply]
Hi @hmottestad,

this is interesting. However, if we send RDF Delta Patches via Ratis we would get almost everything for free.
Ratis manages the Raft Log (with committed RDF Delta Patches) and organizes the updates of the state machine on each node.
Using the NotifyingSail or just a wrapper for the SnapshotSailStore to track the committed updates is also a good idea.

What do you think?

Best regards,
Ken

**Reply by hmottestad** [reply]
That sounds interesting. We don't have an implementation of the RDF Patch format, so we would anyways have to implement that.

I've never used Ratis, so I don't know how much code we would have to write ourselves. For Jena's RDF Delta Patch Log Server we would need to implement the interaction layer with the server, which uses http. Persistence, concurrency control and HA is already part of the server. 

Performance is a major factor. Jena's RDF Delta Patch Log Server uses http for transport and ZooKeeper for HA. I can see that Ratis is very performance focused, but I couldn't find anything about how it compares to ZooKeeper.

**Reply by kenwenzel** [reply]
I also do not yet completely undestand the state machine implementation. Especially, the logic for storing and loading snapshots is not clear to me.

The counter example uses snapshots:
https://github.com/apache/ratis/blob/master/ratis-examples/src/main/java/org/apache/ratis/examples/counter/server/CounterStateMachine.java

The file store example does not:
https://github.com/apache/ratis/blob/master/ratis-examples/src/main/java/org/apache/ratis/examples/counter/server/CounterStateMachine.java

But in the end it is maybe enough to just send the RDF patches as binary data and apply it to the local RDF4J repository.

**Reply by benherber** [reply]
From what I understand having toyed a bit with Ratis is that it follows the ideology of a lot of other RAFT libraries in other languages; handling leader elections, heartbeats, all the protocol things (of course you can plug your own implementations in for specific parts). They seem to have a billion and one configurations to tune exactly how this behaves (e.g., do we want full 'linearizable' behavior or can we loosen that requirement a bit). 

The real parts to implement afaik are 1. Creating a format for serialized log entires (i.e., what is the state machine); 2. how to 'apply' log entries to deterministically recreate a snapshot of the data corresponding to a given log entry (i.e., run the state machine); and 3. how to create and store a snapshot corresponding to a given log entry such that the replicated log can be compacted. 

For a bit more clarity on the protocol part, I found this visualization rather useful: http://thesecretlivesofdata.com/raft/

**Reply by benherber** [reply]
I guess my question would be, to better define 'High-Availability', what guarantees are thought to be appropriate for such a deployment of RDF4J? Is it okay to be purely eventual or causally consistent or is something more strict required. Those are two very different problems to solve as far as I have seen.

**Reply by benherber** [reply]
> Performance is a major factor. Jena's RDF Delta Patch Log Server uses http for transport and ZooKeeper for HA. I can see that Ratis is very performance focused, but I couldn't find anything about how it compares to ZooKeeper.

Ratis and Zookeeper are fairly different in what they are trying to do. RAFT is a trying to come to a consensus on what records are in a WAL for the purposes of applying operations in a deterministic order. It may be more comparable to compare Zookeeper's consensus algorithm [ZAB](https://medium.com/@mehar.chand.cloud/zookeeper-atomic-broadcast-protocol-zab-understanding-its-mechanisms-ffc157c08f5e) to RAFT as they aim to achieve the same end goal. Zookeeper uses this to implement its features like coordination, synchronization primitives, etc. There is[ this really interesting article](https://www.confluent.io/blog/why-replace-zookeeper-with-kafka-raft-the-log-of-all-logs/) from Confluent on them changing their backend to use their own RAFT variant (K-RAFT) instead of Zookeeper. 

Another big thing Ratis likes to tout is that it is not an external service like Zookeeper or Kafka, but rather an embedded library in the server itself which of course comes with a series of benefits and drawbacks.

**Comment by afs** [comment]
(Sorry if I'm butting in here)

Just to be clear; RDF Delta isn't part of Apache Jena. There's no technical barrier but RDF Delta is significantly more user-support-intensive. Not surprising - it is affected by the deployment environment.

Delta could do with a refresh, even a V2. Apache Zookeeper takes quite a lot of effect to deploy and operate. It can not store the patches (size limits).

Apache Ratis for the system lock and metadata would be a good choice.

I'm not clear whether storing bulk data in Ratis for a long-lived deployment is a good idea or not. This might depend if the design is to keep patches "for a long time" (c.f. incremental backup, rebuild new triplestores from a long-term data snapshot) or just until the triplestores have all taken the updates. RDF Delta does not track the state of the front ends.

If patches are outside Ratis, for storing the patches, there are options:

There is a lot to be said in favour of small deployments using a "safe" filesystem for patch storage. The advantage for users is that the deployment is simple to operate. SPARQL is continuously available for query; yes, there is there is a pause on updates if the patch server needs to be bumped. It starts up very fast (1-2s). Only loosing the patches is catastrophic; a blank patch server can discover log state (once) and that takes a little longer to startup.

For a more complex deployment, Apache Pulsar looks interesting. It is designed to be a distributed log. It supports migrating stored data to different storage hierarchies and also ageing off patches. Apache Kafka nowadays can be used in a log-like manner if old patches are migrated away from expensive broker storage.

This still needs Ratis to coordinate writing the log consistently.

Current RDF Delta can use blob stores - that is another operational cost. Pulsar should be a more integrated solution as well as being cloud-neutral.

**Reply by kenwenzel** [reply]
Hi @afs,

thank you for joining the discussion. I also thought about using something like Kafka - and Pulsar looks very interesting.

As far as i understand, Ratis logs can be pruned if they are committed to the state machine. If each state machine has an internal RDF store (TDB or the RDF4J stores) then managing the logs should be not a problem.
In this setup we have the possibility to distribute read requests among all RAFT peers. The downside is that we loose the option to store the log on cheap storage systems like S3.

**Comment by afs** [comment]
Having a machine Ratis in the Ratis membership group could be writing out to the long term log. The members don't have to have the same functionality. Delta minimised implementation on the triplestore process because Zookeeper isn't a library.

@kenwenzel What are your thoughts on recover after a failure?

**Reply by kenwenzel** [reply]
I'm not sure if I understand the architecture right: We have a RAFT group with multiple peers (>= 3) where each is managing an internal RDF database (Here the deltas that led to the content of the RDF database are lost over time if the RAFT log is pruned.). Some peers also write the log entries to a long term storage (Jena's delta log) to keep the full historical state.
Is this correct?

If a failure occurs on a node then I would reconstruct its state by taking the contents of the RDF database from the leader in combination with the (automatically) synchronized RAFT log entries. For transferring the database contents an additional mechanism has to be developed.
Could you imagine that this is working?

**Reply by afs** [reply]
> Could you imagine that this is working?

Yes - with sufficient care during the transfer is linked to the exact RAFT state so it will catchup when it goes live because updates may still be happening (e.g. the cluster was 4, and 1 machine lost).

**Comment by frensjan** [comment]
Has there been any further work done in this space? I could be/get interested in chipping in here.

What would be the desired feature set for HA in RDF4J?

What kind of **transaction support** is desired? Is read-committed isolation okay for interested parties? Are there use cases that would require higher levels of isolation?

Are we only looking at HA here? Or perhaps also **distribution**? Could we mary the two?
We're managing over a 1000 'stores' of varying sizes. Some non-distributed with just a few MB's of metadata. But in total we have quite some terabytes that we currently have in distributed stores (HBase and Elasticsearch).

As for the use of Ratis in this area: I think the data/metadata separation in Ratis built for Hadoop HDDS/Ozone can be used here. This **stream feature** allows users to quite efficiently distribute patches without ending up in the Raft logs. See also [this blog from Cloudera](https://blog.cloudera.com/ozone-write-pipeline-v2-with-ratis-streaming/).

**Reply by fkleedorfer** [reply]
Still very interested. Read committed would be fine. 

Regarding distribution, I am not sure what you mean - not every graph in every store instance?

**Reply by frensjan** [reply]
Sorry, I should have been clearer. I meant sharding / distributed partitioning indeed. Sharding could be by (named) graph, by entity or by statement. We currently partition by (hash of) entity id in HBase. We have an entity centric index in Elasticsearch, so that's naturally distributed at the entity level.

We're not using  named graphs, so I'm not that interested in that. There are some possible use cases in our product, but they would probably be pretty large at least for some of our stores. Sharding/partitioning at the entity level would be interesting. It could have some interesting properties in terms of push-down.

It could be beneficial if we layer sharding on top of a HA solution.

**Reply by fkleedorfer** [reply]
Definitely also interested in sharding, although not as pressing.

**Comment by frensjan** [comment]
For the combination of Raft / Ratis and RDF4J stores (probably any HA solution), it makes sense to create/distribute snapshots. Otherwise the state of a node is essentially a big WAL from which a store needs to be rebuilt on start. Perhaps with some trickery this can be worked around; but e.g. Ratis has a concept of snapshots and it would be nice to "go with it's flow". And for nodes to be added, catch up, be replaced, etc. it's a lot more efficient if they can start from a snapshot, instead of fetching the log and applying that locally.

I've worked with a lot of LSM based databases. For such systems creating a snapshot is often just a matter of creating (hard)links of the files (segments or whatever they're called in the particular system). Is there any such (general) mechanism imaginable for RDF4J stores?

I'm not sure about the mechanics of the native store; I would guess it updates pages in place. I don't believe LMDB has such snapshot capability.

An alternative design is to (shortly) interrupt processing on a node or at least go into read-only mode, and create a copy of the RDF4J on disk data. I believe this is also done in Apache Ozone (for which Ratis was initially developed) with their RocksDB (metadata) databases. Some downsides for that to work:
- followers need to pause processing
- a leader should step down (otherwise the entire 'cluster' is paused)
- it increases the write amplification

**Reply by kenwenzel** [reply]
@frensjan I would also use the snapshot mechanism to allow pruning the Raft log. The NativeStore and LmdbStore both rely on one or few binary files for storing the data and btrees. In the simplest case we could just transfer these binary files to followers. I've already mentioned this in a message above:
https://github.com/eclipse-rdf4j/rdf4j/discussions/5147#discussioncomment-11006692

While the snapshot/data is copied we could store new data in temporary stores (one for added and one for deleted statements) - which is currently also done in MemoryOverflowModels for large transactions - and set the copied store to readonly. After copying is finished we can apply the deltas to the main store.

**Reply by frensjan** [reply]
Thanks @kenwenzel. A bit more involved than hard linking files, but I can see that working conceptually. However, I do see some technical challenges.

I don't see a way yet how this could be done with Ratis as it assumes snapshots are immutable and I have found no way of coordinating snapshot usage (for replication) and log application.

Just creating a snapshot by copying the main store on demand when a node joins could be problematic, depending on the size of the database of course. We have shards that are several 10's of GB's in size. Creating a copy just to install a snapshot on a follower would be troublesome. These things tend to happen also when the proverbial sh*t hits the fan.

I think that creating snapshots of the entire main store often would be very costly from a write amplification point of view. This would be force in the direction of a mutable main store. But this would be at odds with the issue above regarding Ratis and how it handles snapshotting.

Would like to hear your perspective on this.

**Comment by Ostrzyciel** [comment]
Hi! Very interesting conversation here.

I see that most of the discussion here is around the synchronization/leader election/coordination issues – I can't really help with that. But, I'm currently working on a binary implementation of [RDF Patch](https://afs.github.io/rdf-delta/rdf-patch.html) (the serialization part of RDF Delta) based on Jelly: https://github.com/Jelly-RDF/jelly-protobuf/issues/11

The nice part about this is that it should be really, really fast, with built-in compression. You could also use long-running gRPC streams as the transport, which would dramatically cut the latency as compared to HTTP GET ping-pongs.

I will implement this for both Jena and RDF4J with a shared (generically-typed) core.

I'm currently knee-deep in this project, but still at a point where I can make design changes. If you have any insights/requests – please go for it :)

In terms of possible relevance for both Jena and RDF4J, there are 2 things to consider.

- Jena already implements RDF Patch in text and Thrift.  The Thrift format does no compression, though, and in my (first-party) benchmarks it is way slower than Jelly with plain RDF data. So, the main advantage would be possibly better perf.
- I'm currently writing Jelly-JVM in Scala, because it simply saves me a ton of time as compared to Java. I imagine neither Jena nor RDF4J would be ecstatic to add a Scala dependency. But if you were to be interested in using Jelly for HA (and having it as a core-ish module) we probably could rewrite the core of Jelly-JVM in Java. The performance-sensitive parts already are written in Java.

...no idea if this sounds interesting to you, looking forward to your feedback! :)

**Reply by kenwenzel** [reply]
Hi @Ostrzyciel,

that sounds great. If I could have a wish then a plain Java implementation would be really cool.
Including the Scala library is not a show-stopper but it is rather heavy weight.

**Reply by Ostrzyciel** [reply]
Sure. The first implementation will be certainly in Scala, but then we will probably move to pure Java in Jelly-JVM 3.0.0 – this should be in a few months.

**Reply by Ostrzyciel** [reply]
Hey, we have the first version of Jelly-Patch ready :) it's currently a bit under-documented, because we want to gather some early feedback before finalizing it.

Spec: https://w3id.org/jelly/dev/specification/patch/
Implementation (see modules: core-patch, jena-patch, rdf4j-patch):  https://github.com/Jelly-RDF/jelly-jvm

It's all in pure Java and the only dependency is protobuf-java. Scala is now used only for tests :)

Like I said, there is pretty scarce documentation for this for the moment, so if you are interested in better understanding this, just hit me up on my email... I'd be happy to explain. piotr@neverblink.eu

We are also looking for benchmark cases, so if you have some specific workload in mind for this HA feature, we would be more than happy to use it for benchmarks (probably via RiverBench).

**Comment by odysa** [comment]
Would it make sense to explore using a distributed storage layer as Sail’s backend? [TiKV](https://tikv.org), a key-value database built on RAFT, could be a promising candidate.

**Reply by kenwenzel** [reply]
It could be an option. Others are Cassandra/ScyllaDb. The main challenge is that I don't expect the performance of joins to be great. FedX with distributed named graphs could be a better option.

**Reply by odysa** [reply]
> The main challenge is that I don't expect the performance of joins to be great

Is it because of kv sharding?

**Reply by kenwenzel** [reply]
The main challenge are nested loop joins in combination with database latency. In LinkedFactory Pod, which is an RDF-based time-series db, we use batched/vectored evaluation to address higher latencies when reading Parquet files. It could also work with something like TIKV.

**Reply by odysa** [reply]
If I understand correctly, this architecture introduces a compute/storage separation, which means each read requires fetching data from the remote storage layer. Some databases address this by maintaining a local cache, often combined with strategies like prefetching.

**Reply by kenwenzel** [reply]
Yes, exactly. But if you have some time then trying out TIKV may be worth.
For fetching data FedX already provides many optimizations.

**Reply by odysa** [reply]
IMO, at least for our use case, which is read-heavy, a distributed system is overkill. Its downsides outweigh the benefits. A single-leader replication model is sufficient. We can either have the leader stream WAL to the read-only replicas, or adopt a centralized WAL server approach like the RDF Delta Server or Pulsar/Kafka. High availability can be handled by the object storage or file system. （or a Raft-based RDF delta server??)
